### PR TITLE
180982568 weighted apy

### PR DESCRIPTION
**Pull request by maciejkocylapc**
#### What's this PR do?
- Improve apy calculation and add weighted average

#### How should this be manually tested?
- run on stage and check apy

#### What are the relevant tickets?
- [URL to the PivotalTracker ticket](https://www.pivotaltracker.com/story/show/180982568)

#### Task completed checklist:
- [] Is there appropriate test coverage?
- [] Did I take into consideration possible security vulnerabilities?
- [] Are the controllers secure?
- [] Was every important matter documented?

#### Things to watch out for
- [N] Does this PR have any migrations?
- [N] Does this PR add new dependencies?

### CODE REVIEW

File: app/logic/stake_logic.rb [hunk 1]
```diff
@@ -216,54 +216,100 @@ def get_rewards
...
+    end
+  end
+
+  def calculate_apy_for_pools
+    lambda do |p|
+      p.payload[:stake_pools].each do |pool|
+        history_accounts = StakeAccountHistory.where(
+          stake_pool: pool,
+          epoch: p.payload[:previous_epoch].epoch
+        ).select("DISTINCT(stake_pubkey) stake_pubkey, active_stake")
```
Comment (radek-polcode) [inline app/logic/stake_logic.rb:259]: select usually goes first

File: app/logic/stake_logic.rb [hunk 2]
```diff
@@ -216,54 +216,100 @@ def get_rewards
...
+
+        total_stake = 0
+
+        weighted_apy_sum = pool.stake_accounts.inject(0) do |sum, sa|
+          stake = history_accounts.select{ |h| h&.stake_pubkey == sa.stake_pubkey }.first&.active_stake
+
+          if stake && stake > 0
+            total_stake += stake
+            if stake && stake > 0 && sa.apy
```
Comment (radek-polcode) [inline app/logic/stake_logic.rb:268]: `if stake && stake > 0` Isn't it the same condition as above?

File: app/models/stake_account.rb [hunk 3]
```diff
@@ -41,4 +41,11 @@ class StakeAccount < ApplicationRecord
   scope :filter_by_staker, ->(staker) { where('staker LIKE ?', "#{staker}%") }
   scope :filter_by_withdrawer, ->(withdrawer) { where('withdrawer LIKE ?', "#{withdrawer}%") }
   scope :active, ->() { where.not(active_stake: [0, nil]) }
+
+  def history_from_epoch(epoch)
+    StakeAccountHistory.where(
+      stake_pubkey: stake_pubkey,
+      epoch: epoch
+    ).first
```
Comment (radek-polcode) [inline app/models/stake_account.rb:49]: `find_by` instead of `where/first`?

File: test/logic/stake_logic_test.rb [hunk 4]
```diff
@@ -257,11 +257,56 @@ def setup
       }
     )
     p = Pipeline.new(200, @initial_payload)
-                .then(&calculate_apy)
+                .then(&calculate_apy_for_accounts)
     
     acc.reload
-
+    puts p.errors
```
Comment (radek-polcode) [inline test/logic/stake_logic_test.rb:263]: not needed I think

**Review by radek-polcode — CHANGES_REQUESTED** [review]

File: app/logic/stake_logic.rb [hunk 5]
```diff
@@ -216,54 +216,102 @@ def get_rewards
...
-        network: p.payload[:network]
-      ).last
-
-      previous_epoch = EpochWallClock.where(
-        network: p.payload[:network],
-        epoch: last_epoch.epoch - 1
-      ).last
+      last_epoch, previous_epoch = set_epochs(p.payload[:network])
 
       num_of_epochs = 1.year.to_i / (last_epoch.created_at - previous_epoch.created_at).to_i.to_f
```
Comment (radek-polcode) [inline app/logic/stake_logic.rb:231]: This also could be moved to a helper module and tested.

File: test/logic/stake_logic_test.rb [hunk 6]
```diff
@@ -257,11 +257,55 @@ def setup
...
-                .then(&calculate_apy)
+                .then(&calculate_apy_for_accounts)
     
     acc.reload
-
     assert_equal 200, p.code
     assert_equal 12.5502, acc.apy
   end
+
+  test "calculate_apy_for_pools should return correct average apy" do
```
Comment (radek-polcode) [inline test/logic/stake_logic_test.rb:267]: This one test is enough for this feature? There are couple of `if` statements in this method, maybe it's worth to test them too.

**Review by radek-polcode — CHANGES_REQUESTED** [review]
More comment about calculations would be nice for someone who did not create this feature, also some code refactor could help to maintain the code in case of changes of the counting formula, this might be a follow up story.

File: app/logic/stake_logic.rb [hunk 7]
```diff
@@ -216,54 +216,102 @@ def get_rewards
...
-        end
+        credits_diff = reward_with_fee(acc.stake_pool&.manager_fee, rewards[:amount])
 
         next unless credits_diff > 0
-        credits_diff_percent = credits_diff / (rewards[:postBalance] - credits_diff).to_f
 
-        apy = (((1 + credits_diff_percent) ** num_of_epochs) - 1) * 100
-        apy = apy < 100 && apy > 0 ? apy.round(6) : nil
+        apy = calculate_apy(credits_diff, rewards, num_of_epochs)
         acc.update(apy: apy)
```
[the remark this answers is not in the document]
Comment (maciejkocylapc) [reply app/logic/stake_logic.rb:244]: yes, when apy is nil, then we display N / A on the front

**Review by kbiala — COMMENTED** [review]
Please update Metrics Explanation section

**Comment by kbiala** [comment]
Please take a look at conflicts.

File: app/logic/stake_logic.rb [hunk 8]
```diff
@@ -228,19 +229,20 @@ def calculate_apy_for_accounts
...
+        apy = nil  
+        if p.payload[:account_rewards][acc.stake_pubkey]
 
-        rewards = p.payload[:account_rewards][acc.stake_pubkey].symbolize_keys
-        credits_diff = reward_with_fee(acc.stake_pool&.manager_fee, rewards[:amount])
+          rewards = p.payload[:account_rewards][acc.stake_pubkey].symbolize_keys
+          credits_diff = reward_with_fee(acc.stake_pool&.manager_fee, rewards[:amount])
 
-        next unless credits_diff > 0
+          next unless credits_diff > 0
```
Comment (radek-polcode) [inline app/logic/stake_logic.rb:242]: This prevents from update if condition is meet

**Review by radek-polcode — CHANGES_REQUESTED** [review]

File: db/schema.rb [hunk 9]
```diff
@@ -340,6 +340,8 @@
     t.integer "average_lifetime"
     t.float "average_score"
     t.float "average_apy"
+    t.float "withdrawal_fee"
+    t.float "deposit_fee"
```
Comment (radek-polcode) [inline db/schema.rb:344]: Are these two fields added by you?

**Review by radek-polcode — CHANGES_REQUESTED** [review]

File: db/schema.rb (hunk 9 above)
Comment (kbiala) [reply db/schema.rb:344]: There're from different task, merged to master

**Pull request merged**